### PR TITLE
[Feature] select_out_keys

### PR DESCRIPTION
**Pull request by vmoens**
## Description

Implements a `select_out_keys` method for all TensorDictModuleBase subclasses, as an alternative to #349 

## Usage

```
>>> from tensordict import TensorDict
>>> from tensordict.nn import TensorDictModule, TensorDictSequential
>>> import torch
>>> mod = TensorDictModule(lambda x, y: (x+2, y+2), in_keys=["a", "b"], out_keys=["c", "d"])
>>> mod(TensorDict({"a": torch.zeros(()), "b": torch.zeros(())}, []))
TensorDict(
    fields={
        a: Tensor(shape=torch.Size([]), device=cpu, dtype=torch.float32, is_shared=False),
        b: Tensor(shape=torch.Size([]), device=cpu, dtype=torch.float32, is_shared=False),
        c: Tensor(shape=torch.Size([]), device=cpu, dtype=torch.float32, is_shared=False),
        d: Tensor(shape=torch.Size([]), device=cpu, dtype=torch.float32, is_shared=False)},
    batch_size=torch.Size([]),
    device=None,
    is_shared=False)
>>> mod.select_out_keys(["d"])
>>> mod(TensorDict({"a": torch.zeros(()), "b": torch.ones(())}, []))
TensorDict(
    fields={
        a: Tensor(shape=torch.Size([]), device=cpu, dtype=torch.float32, is_shared=False),
        b: Tensor(shape=torch.Size([]), device=cpu, dtype=torch.float32, is_shared=False),
        d: Tensor(shape=torch.Size([]), device=cpu, dtype=torch.float32, is_shared=False)},
    batch_size=torch.Size([]),
    device=None,
    is_shared=False)
```
This works with dispatch too:
```python
>>> mod(torch.zeros(()), torch.zeros(()))
tensor(2.)
```

Can also be reset:
```
>>> mod.reset_out_keys()
>>> mod(TensorDict({"a": torch.zeros(()), "b": torch.ones(())}, []))
TensorDict(
    fields={
        a: Tensor(shape=torch.Size([]), device=cpu, dtype=torch.float32, is_shared=False),
        b: Tensor(shape=torch.Size([]), device=cpu, dtype=torch.float32, is_shared=False),
        c: Tensor(shape=torch.Size([]), device=cpu, dtype=torch.float32, is_shared=False),
        d: Tensor(shape=torch.Size([]), device=cpu, dtype=torch.float32, is_shared=False)},
    batch_size=torch.Size([]),
    device=None,
    is_shared=False)
>>> mod(torch.zeros(()), torch.ones(()))
(tensor(2.), tensor(3.))
```
TODO:
- [ ] docstrings
- [ ] tests
  - TDModule
  - Custom TDModuleBase
  - Tricky signatures 
  - TDSequential
  - Probabilistic modules
  - remove hook

cc @apbard

### CODE REVIEW

File: tensordict/nn/common.py [hunk 1]
```diff
@@ -276,18 +276,106 @@ def wrapper(_self, *args: Any, **kwargs: Any) -> Any:
...
+            return (
+                item
+                for i, item in tensordict_out
+                if module._out_keys[i] in module.out_keys
+            )
+        return tensordict_out.select(
+            *in_keys, *out_keys, inplace=tensordict_out is tensordict_in
+        )
+
+    def _detech_dispatch(self, tensordict_in, in_keys):
```
Comment (tcbegley) [inline tensordict/nn/common.py:319]: Is this meant to be `_detect_dispatch`?

File: tensordict/nn/common.py [hunk 2]
```diff
@@ -276,18 +276,106 @@ def wrapper(_self, *args: Any, **kwargs: Any) -> Any:
...
+            return
+        self._initialized = True
+        self.module = module
+        module.out_keys = self.out_keys
+
+    def __call__(
+        self,
+        module: TensorDictModuleBase,
+        tensordict_in: TensorDictBase,
+        tensordict_out: TensorDictBase,
```
Comment (tcbegley) [inline tensordict/nn/common.py:295]: I think the typing here is confusing as we sometimes expect these to be tuples of tensors right?

**Review by tcbegley — COMMENTED** [review]
This looks really cool!

Should the method be `select_out_keys_` since it modifies the module in place? Or is that a convention that applies to tensors rather than modules?

**Comment by tcbegley** [comment]
One reason we were interested in a change like this was so that we could get control over the `out_keys` of a `TensorDictSequential` which otherwise attempts to infer `out_keys` from the sequence. It would be nice if we could add an argument to the constructor in that case so that we can write something like

```python
module = TensorDictSequential(module1, module2, module3, out_keys=["the_key_i_want"])
```

instead of

```python
module = TensorDictSequential(module1, module2, module3)
module.select_out_keys(["the_key_i_want"])
```

**Comment by vmoens** [comment]
> One reason we were interested in a change like this was so that we could get control over the `out_keys` of a `TensorDictSequential` which otherwise attempts to infer `out_keys` from the sequence. It would be nice if we could add an argument to the constructor in that case so that we can write something like
> 
> ```python
> module = TensorDictSequential(module1, module2, module3, out_keys=["the_key_i_want"])
> ```
> 
> instead of
> 
> ```python
> module = TensorDictSequential(module1, module2, module3)
> module.select_out_keys(["the_key_i_want"])
> ```

While I understand the need for this and how that could be used, I'm reluctant to this.
I will try to summarize here why:
Extra kwarg argument always look like a good idea. Who prefers to call a method to condition the state of the object rather than building it straight away like she wants to?
However, there are a number of arguments against this in this case:
- TensorDictSequential is neat, clear, has no extra kwarg and that makes it compelling to use. The doc of the construction is super simple: just pass a bunch of TDModuleBase instances and you're done. We would add one more level of complexity to this.
- "No big deal, it's just one". I hear you. But it's one too much. The first reason is inheritance: we have quite some subclasses of TensorDictSequential already in tensordict and tochrl. I bet there will be more to come in this library and others. We will ask developers to do one of two things: either ignore the kwarg and break reusability / modularity (ie you won't be able to swap classes easily) or adopt it but then all the `__init__` signature will need to be adapted backward (existing subclasses) and forward (future subclasses). This is an extra burden on all ends:
  - ourselves: we'll need to review each PR with that feature in mind, asking users to add an extra kwarg. If we forget, no class substitution allowed and/or bugs. 
  - our users for the same reason
  Keep in mind also that we're doing this in a context where we already have the solution (proposed by this PR) at hand, the constructor just calls that from what I understood. So it's an extra burden with limited benefit IMO.
- This brings me to the Don't Repeat Yourself principle: we'd be providing two alternative ways of doing the same thing. That isn't great in general: confusing for the users (which one is the best?) and hard to maintain. 

Here are some scenarios where I think we'd be annoyed by this:
- One day, we change `out_keys` in `selected_out_keys` for commodity. We put a deprecation warning in `__init__` then remove it. We're asking all developers that uses `TensorDictSequential` to modify their `__init__` method. On the other hand, changing it in `set_out_keys` only affects us (and maybe users?) although right now you just pass keys as `select_out_keys(key1, key2)` so this scenario does not apply but you got the idea anyway.
- We change something in the docstrings of the behaviour of this key selection (by definition, it could happen but as of now i can't tell you what it is oc). We'll need to chase every subclass and change its docstrings. Again, none of this would be necessary with the method and the method only.

TLDR
The score of TensorDictSequence is to be used as is but also to be subclassed, and as such it is a burden to have extra kwargs in the constructor.
Without considering inheritance, we'd be writing 2 entry points for one functionality, duplicating docstrings etc.
I'm happy to explain further if things are not clear but to be honest I don't think I'm ready to add it to the constructor just yet.

**Review by tcbegley — APPROVED** [review]
Couple of small comments / questions, but otherwise LGTM

File: tensordict/nn/common.py [hunk 3]
```diff
@@ -276,17 +276,291 @@ def wrapper(_self, *args: Any, **kwargs: Any) -> Any:
...
 
-    in_keys: List[str | Tuple[str]]
-    out_keys: List[str | Tuple[str]]
+    def __new__(cls, *args, **kwargs):
+        # check the out_keys and in_keys in the dict
+        in_keys = None
+        if "in_keys" in cls.__dict__:
+            in_keys = cls.__dict__.get("in_keys")
+            # now let's remove it
+            delattr(cls, "in_keys")
```
[the remark this answers is not in the document]
Comment (vmoens) [reply tensordict/nn/common.py:386]: try it out 
Spoiler alert: it raises an exception

File: tensordict/nn/common.py [hunk 4]
```diff
@@ -276,17 +276,291 @@ def wrapper(_self, *args: Any, **kwargs: Any) -> Any:
...
+        out = super().__new__(cls)
+        if in_keys is not None:
+            out.in_keys = in_keys
+        if out_keys is not None:
+            out.out_keys = out_keys
+        return out
+
+    def __init__(self):
+
+        super().__init__()
```
[the remark this answers is not in the document]
Comment (vmoens) [reply tensordict/nn/common.py:401]: I thought it was beautiful but ok
